Differentiate socks5h from socks5 and socks4a from socks4

ijcoe6ru opened the pull request:
In the proxy string, socks5h:// and socks4a:// mean that the hostname is
to be resolved by the socks server.  socks5:// and socks4:// mean the
hostname is to be resolved locally.

Fix #1035

Lukasa commented on test/contrib/test_socks.py:
```diff
@@ -229,6 +229,37 @@ def request_handler(listener):
+            handler = handle_socks5_negotiation(sock, negotiate=False)
+            addr, port = next(handler)
+
+            self.assertIn(addr, ['127.0.0.1', '::1'])
```
You will need to start using `unittest2` in these test cases to use `assertIn`: it's not present on Python 2.6.

ijcoe6ru replied:
Excuse me.  Why are the tests run in Python 2.6 instead of 2.7?

SethMichaelLarson replied:
We have to support Python 2.6 because our upstreams require it.

Lukasa replied:
We don't just run in Python 2.6. We run in Python 2.6, Python 2.7, Python 3.3, Python 3.4, Python 3.5, and PyPy 2, across two different operating systems. =) The tests, and the code, must work on all of those platforms.

ijcoe6ru commented:
I've added tests for doing DNS lookup locally and changed the tests that
use example.com to use socks4a and socks5h.

As of the concern that some code may break, they'll probably be fine
since some well-known software, such as curl, use this convention.

Lukasa commented on test/contrib/test_socks.py:
```diff
@@ -236,7 +236,12 @@ def request_handler(listener):
+            if not addr in ['127.0.0.1', '::1']:
+                raise self.failureException(
+                    "Requested address is %s, 127.0.0.1 or ::1 expected" %
+                    safe_repr(addr))
```
There's no need to do this. You can leave the `self.assertIn` statement and then just edit `dummyserver/testcase.py` to change `import unittest` to

```python
import sys

if sys.version_info >= (2, 7):
    import unittest
else:
    import unittest2 as unittest
```

That should resolve the problem.

ijcoe6ru commented:
Thanks.  I hadn't noticed that unittest2 was used elsewhere, so I thought it would've introduced new dependencies.

Lukasa commented on dummyserver/testcase.py:
```diff
@@ -1,4 +1,8 @@
+if sys.version_info >= (2, 7):
+    import unittest
+else:
+    import unittest2 as unittest
```
So to make flake8 happy, let's move this block down to line 19.

ijcoe6ru commented:
I don't know why the test has failed. Can anyone help me?

Lukasa commented:
It's just a transient failure, I've restarted that test.

SethMichaelLarson commented:
To see details about which tests failed and on what platform, you can click on the Travis details in the required checks box here at the bottom of the PR.  It looks like the test that failed was a randomly failure.

salehe commented:
This is a very useful feature. Looking forward for it to be merged

ijcoe6ru commented:
Why hasn't this been merged yet?

SethMichaelLarson commented:
Huzzah! :D
